### ISSUE TITLE
Release lock on SIGTERM

### ISSUE THREAD

**Issue by MarvinGEA** [liquibase/liquibase#1311]
We're using Liquibase in a Wildfly application server using the liquibase CDI integration inside a Docker container (not sure if the latter fact is relevant for the problem).

If we now terminate the Wildfly while Liquibase is applying our change sets, the changelog lock will not be released. Upon the next start of the Wildfly we're waiting for the lock to be released, which will never happen. Then we require manual intervention to release the lock and restart the Wildfly.

Is there a way to release the lock if the application server receives a SIGTERM signal?

**Comment by MarvinGEA** [comment]
This seems to be a problem with Spring Boot too: https://stackoverflow.com/questions/55954588/ensuring-spring-boot-and-liquibase-receive-and-handle-sigterm

**Comment by molivasdat** [comment]
Hi @MarvinGEA Thanks for bringing this issue to our attention.  There is currently not a way to release the changelog lock when the application receives a SIGTERM.  If you would like to suggest a way via a PR with code and create a fix, we can review it for inclusion into the code.

**Comment by MarvinGEA** [comment]
We already tried to fix it by adding a PreDestroy hook to the `CDILiquibase` class:

```
    @PreDestroy
    public void releaseLock() {
        log.info("Releasing Liquibase lock.");
        // In case of a shutdown of the Wildfly during the execution of liquibase changesets the liquibase lock may not have
        // been released.
        Database database = null;
        try {
            database = createDatabase(dataSource.getConnection());
            LockService lockService = LockServiceFactory.getInstance().getLockService(database);
            lockService.releaseLock();
        } catch (Exception e) {
            log.severe("Error releasing Liquibase lock", e);
        }
    }
```

But the CDI container (in this case Wildfly) doesn't execute that hook when shutting down. Unfortunately our CDI/Java EE knowledge is not that advanced that we could figure our why it's not executed. Does anyone have an idea?

**Comment by bayetovsky** [comment]
Do you have any shutdown delay to allow servlets to unregister? From springboot/tomcat perspective i'm thinking on something similar to unloadDelay.

Did a quick check on Wildfly's docs and saw this which may give you some additional time to allow servlets to unload -  "org.wildfly.sigterm.suspend.timeout". Can you have a go at it @MarvinGEA ?

**Comment by stoffel2107** [comment]
Hey,

I ran into the same problem using SpringBoot and Liquibase.

The problem here is that the spring-boot shutdown hook that would allow a graceful shutdown (unregistering beans) is added AFTER liquibase has finished...

In simpler words: During the liquibase run, there are only kills... No graceful terminations.

I currently can work around this by simply registering a manual Shutdown Hook before starting spring. If triggered it scans all threads if any thread contains a liquibase stack trace if so it waits and checks again -> until the liquibase instance is done and the shutdown can be executed.

But it would be nicer if either liquibase or spring-boot takes care of this.

The question is: Can we safely exit liquibase during an update? And how? A rollback/or mid update stop would have to be used?!?

I think liquibase should have their own shutdown hook to make sure everything runs smoothly and spring should only have to adapt to this.

**Comment by molivasdat** [comment]
Hi @nvoxland @mariochampion. Do you have any suggestions for @stoffel2107? Also @stoffel2107  can you give more perspective on why you needing to stop an update midway through the application startup so that we can give better advice?

**Comment by bayetovsky** [comment]
I believe the point isn't that you "need" to stop an update midway through the application start intentionally. The case we ran into when I raised it into stackoverflow was in situation in which running a service with Liquibase in a platform such as AWS can lead you to situation in which the startup time exceeds the expected startup time and the container will be terminated as being unhealthy - that will lead to the abrupt interruption of Liquibase and the lock will not be released.

After that and by having that lock in place you will then have an infinite cycle of containers failing as new ones keep getting spawned - which can lead to service degradation/outage as your scaling ability will be affected.

Sample timeline:
- Container 1 starts
- Container 1 experiences delays because of slower dependencies or whatever (such as an APM agent)
- Container 1 starts running liquibase scripts (gets lock updated in DB)
- Container 1 gets marked as unhealthy due to "slow start" and is terminated (without releasing lock)
- Container 2 starts
- Container 2 fails to get the lock and waits until it's marked as "unhealthy" and is terminated
- Container 3 starts
- Container 2 fails to get the lock and waits until it's marked as "unhealthy" and is terminated
- Cycle repeats endlessly until manual intervention is done to clear the lock


@molivasdat Hopefully this helps you get a better perspective of the problem and why the update can be interrupted (and needs to be handled)

**Comment by nvoxland** [comment]
Yes, that is a limitation of the current databasechangeloglock setup. We use a record saved to that table to prevent concurrent attempts to do a liquibase update, but there is no database-standard way of saying "undo this commit when my connection is closed". This is normally fine, except when the VM exists during an update like you see. Then it gets stuck. The rise of AWS-style "kill the process if it takes too long to start up" checks has definitely made the problem worse.

I've tried adding JVM shutdown hooks etc. but those seem to run late enough and/or inconsistently enough that I've not tried to rely on them. 

We have open issues to address it in database-specific manners (See #872) but it is going to need to be some database-specific changes, so we're working on figuring out the best way to handle it in as consistent a manner as possible.

In the mean time, depending on your setup you could use the https://github.com/liquibase/liquibase-nochangeloglock extension to completely disable the databasechangeloglock lock logic. Or use your own custom LockService implementation with whatever custom coordination makes sense for you. WARNING: if you do this, it's up to you to coordinate your systems so only one update is performed at a time, such as by putting it in an "init" container or something.

**Comment by bayetovsky** [comment]
@nvoxland It may still be worth having a go at the suggestion/idea around using Shutdown hooks provided as the mentioned "AWS-style"  still sends signals that could be handled (if i'm not mistaken it's not a SIGKILL that is issued to the process).

That in theory could allow for some reaction to the unexpected closure and release of the lock mechanism in databasechangeloglock.

**Comment by molivasdat** [comment]
Hi @bayetovsky Thanks for the use case.  This is exactly the type of information that I was looking to have in the issue statement.

**Comment by stoffel2107** [comment]
Hey, @bayetovsky's case is similar to mine.

I deploy my applications using Chef and a handmade tool.. During deployment, it gets restarted twice with a SIGTERM.

My workaround is basically just waiting for liquibase to finish.. In easier terms => I could simply make a `join` on the main thread in this shutdown hook, this would cause the shutdown to wait for SpringBoot to initialize.. After spring boots initialization, their shutdown hook would trigger the complete shutdown (Remove the shutdown hook after spring boot-start).

The only thing that would be a nice to have is simply a functionality to stop it mid update and therefore liquibase to take care of everything.

**Comment by nvoxland** [comment]
Yes, #1453 would fix the lock problem you are hitting but it would still be nice to have a more graceful handling of SIGTERM vs. SIGKILL. 

We'll handle the lock issue first, since I still don't trust all setups will not kill off the process before our cleanup work is done so having the better SIGKILL recovery is the safer route. But then better handling of the SIGTERM can be handled. 

We're planning a refactoring of the higher-level command/operation classes, and a cleaner termination handling is probably best done as part of that.

**Comment by blagerweij** [comment]
I stumbled upon this Liquibase extension which uses a RDBMS lock function instead of a table. The advantage is that the RDBMS lock will be released automatically when the database connection is closed (i.e. when the container is stopped unexpectedly).
See [https://github.com/blagerweij/liquibase-sessionlock](https://github.com/blagerweij/liquibase-sessionlock)
Disclaimer: I'm not the original author, but I forked the project and released to Maven central.

**Comment by keeely** [comment]
Flyway doesn't seem to have this problem, so for any new work we're using it instead.  From the FAQ at:
https://flywaydb.org/documentation/learnmore/faq
It says:  _Flyway uses the locking technology of your database to coordinate multiple nodes. This ensures that even if multiple instances of your application attempt to migrate the database at the same time, it still works. Cluster configurations are fully supported._
I don't have much knowledge in this area, so a naive question might be could Liquibase do the same thing?  It appears to work without a hitch in Flyway, and without any of the workarounds described here.

**Comment by broodjetom** [comment]
I have added the following to my code:

```java
Runtime.getRuntime().addShutdownHook(new Thread(() -> {
    try {
        liquibase.forceReleaseLocks();
    } catch (LiquibaseException e) {
        throw new RuntimeException(e);
    }
}));
```

My expection would be that this releases the lock in case a SIGTERM is received.

To test this, I need to run migrations and send a SIGTERM to the application. Since locally the migrations are generally fast, I needed to enforce a long running operation as part of the migrations. I found two ways:

1. Make use of the `executeCommand` operation to then do some kind of sleeping (I used Pythons `time.sleep(60)`).

   ```xml
   <executeCommand timeout="60s" executable="python">
     <arg value="-c"/>
     <arg value="import time; print('Before'); time.sleep(10); print('After');"/>
   </executeCommand>
   ```
2. Make use of the `sql` operation to then do some kind of sleeping (I used PostgreSQL's `pg_sleep(60)`).

   ```xml
   <sql>SELECT pg_sleep(60)</sql>
   ```

With these two operations I did some testing:

1. During the sleep of the operation, when a SIGTERM was received, the `executeCommand` operation was cancelled/stopped, all changes in the current changeset were rolled back, and the lock was released. (an actual graceful shutdown!)
2. During the sleep of the operation, when a SIGTERM was received, the `sql` operation kept running, and actually the changeset was completed, and even changesets coming after it were completed. Only after all changesets were done, the process would end and the lock was removed (though, not sure if this is because of the shutdown hook or just the fact that Liquibase was done executing the migrations).

What is the difference between these two operations in Liquibase, and why is it able to "quit" the `executeCommand` and stop running migration, while it is not able to stop the `sql` operation or at least stop the current and subsequent changeset from running?

**Comment by thecoolestcoderrushil** [comment]
a changelog lock that is never released when the container is terminated is a painful failure, since the next container fails to get the lock and the cycle repeats until manual intervention. rehearsing the migration on a production-shaped twin first would surface the lock problem before deploy day. antifailure.dev (https://antifailure.dev) builds disposable database twins for migration rehearsal and is open source and free to try, with demos available for teams that want one.
